README.md
# alignall
# alignall 
## currently is a re-packaged version of alignall_SNP.P.v11.py, 
## 		which attempts to organize and compartmentalize the 
## 		capabilities of the original script


## 190723 UPDATE
### - began testing script on past ATAC projects 
### - script currently broken down into: alignall.py, alignall_ATAC.py
### 		align_sc.py (under construction), alignall_ChIP.py, alignall_snp.py
 




CHANGELOG.md
# Change Log for All files in Alignall repo 
## Doesn't need to be edited every commit, but keeping track of changes when they are made is helpful. 

#### alignall.py 
<date> | <change_in_detail>
------------------------------
180426 | Added exRNA options in steps and in main before setup(), as well as in setup().  Added exRNA code in align() under miRNA code.  exRNA won't use any hairpin files.  Will change for miRNA in future update.
180618 | Minor change: added library creation protocols in help notes
180924 | Major organization
182031 | Added reference to parseST() (function to make simple text files by parsing sample tracker)
190103 | Added version number, manually, to beginning, and option to print version and exit; also somewhat recently added random barcodes option for small RNA, which still needs a lot of work (generally just need to improve small RNA stuff)
190207 | Added import calls to all files that will eventually be used for SNP, ATAC, and ChIP analyses -> enough to change version to 12 (also going to be adding miRNA to this, as well as random barcode functionality)
190213 | Added parallel cutadapt prototype, added and cleaned up align() with trimming and cleaner miRNA parts, and updated version to 13 (atac part worked on myriam's, but haven't added accessibility and not sure if whole thing will work)
190321 | Changed the settings print out to print from variable which will also print to a "runLog.txt" file directly after setup
190424 | Added S_AUREUS genome detection, but only starpath, gtf, and refflat (and refflat might not even be working)
190430 | Added another S_Aureus genome, and a user-question to differentiate between S_AUREUS genomes
190509 | Added logfile output for kallisto run (outputs both stderr and stdout to log.txt)
190510 | Added hostname determination so we can establish the filesystem after knowing which machine we're on
190522 | Fixed hostname determination so it's simpler also added functionality if only want to do counts and several alignmentdirs exist
190603 | Created the new script (this one) to package up necessary components
190724 | Now works for single end rna-seq alignnment, continuing testing and fixing for atac, chip, snp, etc...
190726 | Added combineLanes, a function which combines files of the different lane, same sample
191014 | Python3-ized much of this script and added multiprocessing of combining lanes.
191205 | Been adding functionality for a lot of the callsnps option stuff, and added option countmirna for only doing mirna counts 
191206 | Changed strand options from 1,2,3 to 0,1,2
191211 | Standard for cutadapt: f we already trimmed, then files should be in fastqDir and program will detect if they're already trimmed so won't run cutadapt again
200127 | Polishing the variant calling pipeline (uses return values from multiprocessing pools to determine success/failure of command)
200420 | No longer going to work on smallRNA seq in this pipeline because docker image (genboree) is working on Linux machine for this
200512 | Adding dbsnp for mouse 
200522 | Added whole exonic file for mouse
200605 | Added an md5 generation step in case people wanted to submit data to GEO 
200707 | Added ability to select "OTHER" for genome, so that we can do wacky mouse-rat-human hybrids without adding to this script 
201026 | Added a function that shows all the commands I used to make a *working* rRNA intervals file, with the hopes it can eventually be applied to other genomes




TEST.py
# script designed to test variations of alignall.py

#tests: 
#1. rna seq
#2. variant calling
#3. chip with input
#5. atac (behaves the same as chip without input)
#6. single cell
#7. single cell variant calling (demuxlet & freemuxlet)
#8. rna seq -> variant calling -> single cell -> demux (if time)
#9. adapter clipping
#10. kallisto (transcriptome alignment)
#11. picard and qc 

#features:
# - timing (need a @time decorator function to time each test)
# - logs (write a central log file for general processes and an individual one for each process run)
# - more tests (there's probably lots of edge cases that won't get tested by above tests)



REFDIR = "" #root directory of all above test files




align_sc.py
#!/usr/bin/env python3 

import subprocess, os, sys
import multiprocessing 
import scanpy as sc
import pandas as pd
from matplotlib import rcParams

def getOutfileName(name):
	#looking in reverse
	for i in range(len(name)-2):
		if name[-i-1:-i-4:-1]=="S__":
			return name[:-i-3] + "out"
			
	for i in range(len(name)-1):
		if name[-i-1:-i-3:-1]=="__":
			return name[:-i-2] + "out"
		elif name[-i-1:-i-3:-1]=="S_":
			return name[:-i-2] + "out"
	return "nosplitchar"

def align_sc(targets):
	for i in targets:
		if i.find("_R1") > 1:
			outfile = getOutfileName(i)
			if outfile == "nosplitchar":
				print "Sample name formatting is required! (samples must contain \"__\" or \"_S\"), exiting..."
				sys.exit()
			try:
				os.mkdir(alignmentDir+"/"+outfile)
				# if aligner == "kallisto":
				# 	os.mkdir("KallistoOuts/"+outfile+"/logs.txt") #don't need this anymore -- just going to need a single alignment dir named for whatever aligner beig used
			except OSError:
				pass

	for target in targets:
		if target.endswith('.fastq.gz'):
			rfc = " --readFilesCommand "+unzipInstruct
		else:
			rfc = ""

		#handle paired 
		subprocess.call("STAR --runThreadN 12 --genomeDir /data/genomes/STAR_HUMAN/   --readFilesIn CLH8_fastq/CLH8_S5_L001_R2_001.fastq.gz   CLH8_fastq/CLH8_S5_L001_R1_001.fastq.gz\
		--readFilesCommand zcat --outFileNamePrefix CLH8-starSolo/   --outFilterType BySJout   --outFilterMultimapNmax 1   --outFilterMismatchNmax 999   --outFilterMismatchNoverLmax 0.04\
		--alignIntronMin 20   --alignIntronMax 1000000   --alignMatesGapMax 1000000   --alignSJDBoverhangMin 1   --soloType Droplet   --soloUMIlen 12\
		--soloCBwhitelist "+whitelistBarcodes)


		STAR --runThreadN 12 --genomeDir /data/genomes/STARHUMAN_272b/ --readFilesIn /data/projects/Jyoti_scRNA/fastq/gexl/061919Fibroid_GEXL_fastqs/061919Fibroid_GEXL_S24_L001_R1_001.fastq.gz /data/projects/Jyoti_scRNA/fastq/gexl/061919Fibroid_GEXL_fastqs/061919Fibroid_GEXL_S24_L001_R2_001.fastq.gz --readFilesCommand zcat --outFileNamePrefix Fibroid_061919/   --outFilterType BySJout   --outFilterMultimapNmax 1   --outFilterMismatchNmax 999   --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20   --alignIntronMax 1000000   --alignMatesGapMax 1000000   --alignSJDBoverhangMin 1   --soloType Droplet   --soloUMIlen 12 --soloCBwhitelist /data/genomes/barcodes/3M-february-2018.txt



#get sequence QC data
#align sc data
	#do cliipping if necessary
	#get alignment QC and logs
#begin SC analysis 
	



alignall_sc2.py
#!/usr/bin/env python3.6

import subprocess, os, string #command line calls, path showing, string things?
import multiprocessing #for parallelizing setuff
import csv #not sure where this is used
import math #not sure where this is used
import sys #program control
import argparse #for easy argument input
import re #for regex searching
import gzip #used for parsing fastq.gz files
import socket #this is used for determining the hostname
import json #this is used for counts (parsing kallisto run_info.json
from builtins import input
from datetime import datetime
from functools import partial
from collections import OrderedDict
import slack

exDir = os.getcwd()


VERSION = 1

npc = multiprocessing.cpu_count()
npc = int(npc*.40)
cur_time = datetime.now().strftime("%y%m%d_%H%M%S")

warningContinue = """WARNING: No arguments were specified, running with the following defaults:
genome=MOUSE
aligner=STAR
fastqDir=fastq
strand=3 (2nd strand)
steps=alignAndCount
indel=no
threads=40pct available
Are you sure you want to continue? (Y/N): """

parser = argparse.ArgumentParser(description="Align fastq reads.")
parser.add_argument("-g", "--genome",    default="MOUSE", type=str,                       help="-g [--genome]       Chose from STAR indexes in /Volumes/Projects/Genomes/ \n Examples: MOUSE, HUMAN (default: MOUSE")
parser.add_argument("-f", "--fastqDir",  default="fastq", type=str,                       help="-f [--fastqDir]     This is the directory in which to find the fastq files to align.\n default: ./fastq")
parser.add_argument("-p", "--steps",     default="all",   type=str, help="-p [--steps] These are the steps in the pipeline to run.\nChoices include all, alignAndQC, align, qcOnly, cluster, and projection.\n(default: alignAndQC)")
parser.add_argument("-t", "--threads",   default=npc,       type=int,                       help="-t [--threads]      Run this many samples concurrently.\n(default: " + str(npc) + ")")
parser.add_argument("--version",         action="store_true",                             help="[--version]         prints the version of the script and then exits") #will be very helpful for when I don't remember which version was used for a project, and what that means for the settings/changes
args = parser.parse_args()

if args.version == True:
        print("alignall main script, version: "+ str(VERSION))
        print("exiting...")
        sys.exit()

genome = args.genome
fastqDir = os.path.abspath(args.fastqDir)
steps = args.steps.lower()
threads = args.threads



runSettings = """########### RUNNING WITH OPTIONS: ###########
######### GENOME:\t\t"""+str(genome)+"""
######### FASTQDIR:\t\t"""+str(fastqDir)+"""
######### STEPS:\t\t"""+str(steps)+"""
######### THREADS:\t\t"""+str(threads)+'\n'
print(runSettings)

if len(sys.argv[1:]) == 0:
        usercontinue = input(warningContinue).upper()
        if usercontinue == 'Y':
                print("Running with defaults...")
        else:
                print("Exiting...")
                sys.exit()

print('Directory of fastq files is ' + fastqDir)
try:
        os.path.isdir(fastqDir)
except:
        print("Fastq directory does not exist. Check your files.")
        sys.exit()
print(steps)
if steps not in ["all", "alignandqc", "align", "cluster", "projection", "all", "qconly"]: # new idea for feature: definition that checks mismatch of 2 betwween entered step and list values, if close, askss user if they wante to continue with suggested step.
        print("The chosen step is not currently supported. Please specify the correct option for -p [--steps].")
        print("Options include all, alignAndQC, align, cluster, projection, qcOnly")
        sys.exit()

if genome == "HUMAN":
        ref = "Homo_sapiens.GRCh38.dna.primary_assembly.fa"
        refIdx = "Homo_sapiens.GRCh38.dna.primary_assembly.fa.fai"
        gtf = "Homo_sapiens.GRCh38.78.gtf"
        refFlat = "humanGRCh38.78_mRNA_refFlat.txt"
        starpath = "STARHUMAN"
        star_miRNApath = "STARsRNA/human/human_miRNA"
        star_tRNApath = "STARsRNA/human/human_tRNA"
        star_yRNApath = "STARsRNA/human/human_yRNA"
        miRNA_genelist = "STARsRNA/human/human_files/matureHUMAN_dna_genes.txt"
        kallistoIDX = "Kallisto_HG38/Homo_sapiens.GRCh38.ncrna.cdna.idx"
        kalmiRNAMature = "miRNA/humanKalisto/matureIndex"
        macsGenome = "hs" #need to skip this in the prefix adding loop
        exonBed = "Homo_sapiens.GRCh38.78.wholeExonicRegions.v3.bed"
        indels = "Mills_and_1000G_gold_standard.indels.hg38.Ens.vcf.gz"
        dbsnpHighCon = "1000G_phase1.snps.high_confidence.hg38.Ens.vcf.gz"
        dbsnp = "Homo_sapiens.fix2.vcf.gz"
        GenomeWindows = "Homo_sapiens.GRCh38.dna_sm.primary_assembly.100bp.bed"

elif genome == "MOUSE":
        ref = "Mus_musculus.GRCm38.dna.primary_assembly.fa"
        refIdx = "Mus_musculus.GRCm38.dna.primary_assembly.fa.fai"
        gtf = "Mus_musculus.GRCm38.78.gtf"
        refFlat = "mouseGRCm38.78_mRNA_refFlat.txt"
        starpath = "STARMOUSE"
        star_miRNApath = "STARsRNA/mouse/mouse_miRNA"
        star_tRNApath = "STARsRNA/mouse/mouse_tRNA"
        star_yRNApath = "STARsRNA/mouse/mouse_yRNA"
        miRNA_genelist = "STARsRNA/mouse/mouseFiles/mouse_miRNA_dna_genes.txt"
        kallistoIDX = "Kallisto_MM38/Mus_musculus.GRCm38.ncrna.cdna.idx"
        kalmiRNAMature = "miRNA/mouseKalisto/matureIndex"
        macsGenome = "mm"
        exonBed = ""
        indels = ""
        dbsnpHighCon = ""
        dbsnp = ""
        GenomeWindows = ""

else:
        print("#### Warning: May have difficulty with scRNA or SNP calling  ####")
        print("#### Genome is not currently fully supported. Good Luck!     ####")

fs_noprefix = [ref, refIdx, gtf, refFlat, starpath, star_miRNApath, star_tRNApath,
                star_yRNApath, miRNA_genelist, kallistoIDX, kalmiRNAMature,
                macsGenome, exonBed, indels, dbsnpHighCon, dbsnp, GenomeWindows]
if socket.gethostname() == "functionalgenomics":
        prefix = "/data/genomes/"
        [ref, refIdx, gtf, refFlat, starpath, star_miRNApath, star_tRNApath, star_yRNApath,
        miRNA_genelist, kallistoIDX, kalmiRNAMature, macsGenome,
        exonBed, indels, dbsnpHighCon, dbsnp, GenomeWindows] = [prefix+x for x in fs_noprefix]
        macsGenome = macsGenome.split("/")[-1]
        unzipInstruct = "zcat"
        whitelistBarcodes = "/data/barcodes/3M-february-2018.txt"

elif socket.gethostname() == "LBCJoshPollack2.ucsf.edu":
        prefix = "/Volumes/Pegasus/Genomes/"
        [ref, refIdx, gtf, refFlat, starpath, star_miRNApath,
        star_tRNApath, star_yRNApath, miRNA_genelist, kallistoIDX, kalmiRNAMature,
        macsGenome, exonBed, indels, dbsnpHighCon, dbsnp, GenomeWindows] = [prefix+x for x in fs_noprefix]
        macsGenome = macsGenome.split("/")[-1]
        unzipInstruct = "gzcat"
else:
        print("not sure what server this is...")
        sys.exit()

alignmentDir = "Alignments/STARsolo"+cur_time
alignmentQCDir = "AlignmentsQC/STARsolo"+cur_time




def main(argv):

    global alignmentDir, alignmentQCDir, threads, exDir

    try: #assign paired/single before checking lane #
            targets = os.listdir(fastqDir)
            targets = [i for i in targets if "fastq.gz" in i] 
            targetsFullPath = [fastqDir+"/"+x for x in targets]
    except OSError:
            print("didn't find a fastq directory...")
            print("cur fastqDir: " + fastqDir)
            sys.exit()
    targetsShort = []
    for f in targets:
        splitf = f.split("_")
        splitf = splitf[:-4]
        targetsFinal = ""
        for l in splitf:
            targetsFinal = targetsFinal+l+"_"
        targetsFinal = targetsFinal[:-1]
        targetsShort.append(targetsFinal)

    targetsShort = set(targetsShort)
    targetsCommaSep = []
    for h in targetsShort:
        fastqMatch = [i for i in targetsFullPath if h in i]
        targetsR1 = [j for j in fastqMatch if "_R1" in j]
        targetsR2 = [k for k in fastqMatch if "_R2" in k]
        targetsR1.sort()
        targetsR2.sort()
        R1s = ""
        R2s = ""
        indTarget = []
        for l in targetsR1:
            R1s = R1s+l+","
        for m in targetsR2:
        	R2s = R2s+m+","
        R1s = R1s[:-1]
        R2s = R2s[:-1]
        indTarget = [R1s,R2s]
        targetsCommaSep.append(indTarget)
    try:
            os.mkdir("AlignAllOut")
    except OSError:
            print("AlignAllOut already exists, starting here.")
            pass
    os.chdir("AlignAllOut") # make the other directory changing a lot easier
    
    runLog = open("runLog.txt", 'w')
    runLog.write(runSettings)

    #make the rest of the important dirs
    for d in ["Alignments", "AlignmentsQC", "SNPcalling", "Peakcalling",
            "SeqQC", "SeqQC/fastqc", "trimdfastqs", "cutadaptLogs"]:
            try:
                    os.mkdir(d)
            except OSError:
                    print("didn't make dir: "+d)
                    pass
    if steps in ["alignandqc", "align", "all"]:
        try:
            os.mkdir(alignmentDir)
            os.mkdir(alignmentQCDir)
        except:
            pass
        align_sc(targetsCommaSep)
        if steps in ["alignandqc", "all"]:
            alignDirs = os.listdir(alignmentDir)
            alignDirs = [i for i in alignDirs if "out" in i]
            alignDirs = [alignmentDir + "/" + i for i in alignDirs]
            print(alignDirs)
            print(alignDirs)
            pool = multiprocessing.Pool(int(threads/2))
            pool.map(cellFilter, alignDirs)
            pool.close()
            pool.join()

    if steps in ["qconly"]:
        try:
            print(os.getcwd())
            if len(os.listdir("Alignments")) > 1:
                user_continue = ""
                print("Multiple alignments have been run,\nwhich alignment dir do you want to perform QC on?\n["+" ".join(os.listdir("Alignments"))+"] ")
                while user_continue not in os.listdir("Alignments"):
                    user_continue = input("\nAwaiting response, ^C to exit: ")
                if user_continue in os.listdir("Alignments"):
                    alignmentDir = "Alignments/"+user_continue
                if user_continue in os.listdir("AlignmentsQC"):
                    alignmentQCDir = "AlignmentsQC/"+user_continue
            alignDirs = os.listdir(alignmentDir)
            alignDirs = [i for i in alignDirs if "out" in i]

            alignDirs = [alignmentDir + "/" + i for i in alignDirs]
            print(alignDirs)
            pool = multiprocessing.Pool(int(threads/2))
            pool.map(cellFilter, alignDirs)
            pool.close()
            pool.join()

        except OSError:
                print("couldn't find the Alignments dir...exiting...")
                sys.exit()





def align_sc(targetsI):
    for i in targetsI:
        if i[0].find("_R1") > 1:
            outfile = getOutfileName(i)
            cwd = os.getcwd()
            if outfile == "nosplitchar":
                print("Sample name formatting is required! (samples must contain _L00), exiting...")
                sys.exit()
            try:
                os.mkdir(alignmentDir+"/"+outfile)
                # if aligner == "kallisto":
                #     os.mkdir("KallistoOuts/"+outfile+"/logs.txt") #don't need this anymore -- just going to need a single alignment dir named for whatever aligner beig used
            except OSError:
                print("couldn't created " + alignmentDir+"/"+outfile)
                print("continuing")
                pass
        #print("running : " + str(target))
        R1fastq = i[0]
        R2fastq = i[1]
        print(outfile)
        print("running: \n" + R2fastq + "\n" + R1fastq + "\nGoing in : " + outfile)
        if R1fastq.endswith('.fastq.gz'):
            rfc = " --readFilesCommand "+unzipInstruct
        else:
            rfc = ""
        #handle paired 
        subprocess.call("STAR --runThreadN "+ str(threads) + " " + "--genomeDir "+starpath+rfc+"  --readFilesIn "+ R2fastq + " " + R1fastq + " \
         --outFileNamePrefix "+alignmentDir+"/"+outfile+"/" + " --outFilterType BySJout   --outFilterMultimapNmax 1   --outFilterMismatchNmax 999   --outFilterMismatchNoverLmax 0.04\
        --alignIntronMin 20   --alignIntronMax 1000000   --alignMatesGapMax 1000000   --alignSJDBoverhangMin 1   --soloType Droplet   --soloUMIlen 12\
        --clip3pAdapterSeq GATCGGAAGAGCACACGTCTGAACTCCAGTCAC --outSAMtype BAM SortedByCoordinate --soloCBwhitelist "+whitelistBarcodes ,shell=True)

def cellFilter(alignDir):
	subprocess.call("Rscript --vanilla " + exDir + "/cellFiltering.R " + alignDir + "/Solo.out",shell=True)


def getOutfileName(name):
    #looking in reverse
    namei = name[0].split(",")[0]
    namei = namei.split("/")[-1]
    if len(namei.split("_L00")) > 1:
        namei = namei.split("_")
        namei = namei[:-4]
        finalName = ""
        for l in namei:
            finalName = finalName+l+"_"
        finalName = finalName[:-1]
        return  finalName + "out"
    else:
         return "nosplitchar"

if __name__ == "__main__":
        main(sys.argv[1:])
